File: 28-find-the-index-of-the-first-occurrence-in-a-string/find-the-index-of-the-first-occurrence-in-a-string.swift
class Solution {

    func hasSubstring(
        _ substring: String,
        in string: String, 
        from startIndex: String.Index
    ) -> Bool {

        var index = startIndex

        for substringCharacter in substring {

            if index == string.endIndex {
                return false
            }

            if substringCharacter != string[index] {
                return false
            }

            index = string.index(after: index)
        }

        return true
    }

    func strStr(_ haystack: String, _ needle: String) -> Int {

        for startIndex in haystack.indices {

            if hasSubstring(needle, in: haystack, from: startIndex) {
                return haystack.distance(from: haystack.startIndex, to: startIndex)
            }
        }

        return -1
    }

    func strStr_VersionWithStandartFunctions(_ haystack: String, _ needle: String) -> Int {
        let range = haystack.range(of: needle)
        
        guard let lowerBound = range?.lowerBound else {
            return -1
        }
        
        return haystack.distance(from: haystack.startIndex, to: lowerBound)
    }
}

File: 28-find-the-index-of-the-first-occurrence-in-a-string/find-the-index-of-the-first-occurrence-in-a-string.java
class Solution {
    public int strStr(String haystack, String needle) {
        
        for (int i = 0; i < haystack.length(); i++) {
            
            if (haystack.length() - i < needle.length()) {
                return -1;
            }
            
            var isEqual = true;
            
            for (int j = 0; j < needle.length(); j++) {
                
                if (haystack.charAt(i + j) != needle.charAt(j)) {
                    isEqual = false;
                    break;
                }
            }
            
            if (isEqual) {
                return i;
            }
        }
        
        return -1;
    }
}
